airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/DebeziumMessageProducer.java

```diff
@@ -0,0 +1,138 @@
+import io.airbyte.cdk.integrations.source.relationaldb.state.SourceStateMessageProducer;
+import io.airbyte.protocol.models.v0.AirbyteMessage;
+import io.airbyte.protocol.models.v0.AirbyteStateMessage;
+import io.airbyte.protocol.models.v0.ConfiguredAirbyteStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DebeziumMessageProducer<T> implements SourceStateMessageProducer<ChangeEventWithMetadata> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DebeziumMessageProducer.class);
+
+  private final CdcStateHandler cdcStateHandler;
+
+  /**
+   * `checkpointOffsetToSend` is used as temporal storage for the offset that we want to send as
+   * message. As Debezium is reading records faster that we process them, if we try to send
+   * `offsetManger.read()` offset, it is possible that the state is behind the record we are currently
+   * propagating. To avoid that, we store the offset as soon as we reach the checkpoint threshold
+   * (time or records) and we wait to send it until we are sure that the record we are processing is
+   * behind the offset to be sent.
+   */
+  private final HashMap<String, String> checkpointOffsetToSend = new HashMap<>();
+
+  /**
+   * `previousCheckpointOffset` is used to make sure we don't send duplicated states with the same
+   * offset. Is it possible that the offset Debezium report doesn't move for a period of time, and if
+   * we just rely on the `offsetManger.read()`, there is a chance to sent duplicate states, generating
+   * an unneeded usage of networking and processing.
+   */
+  private final HashMap<String, String> initialOffset, previousCheckpointOffset;
+  private final AirbyteFileOffsetBackingStore offsetManager;
+  private final CdcTargetPosition<T> targetPosition;
+  private final Optional<AirbyteSchemaHistoryStorage> schemaHistoryManager;
+
+  private boolean shouldEmitStateMessage = false;
+
+  private final DebeziumEventConverter eventConverter;
+
+  public DebeziumMessageProducer(
+                                 final CdcStateHandler cdcStateHandler,
+                                 final CdcTargetPosition targetPosition,
+                                 final DebeziumEventConverter eventConverter,
+                                 final AirbyteFileOffsetBackingStore offsetManager,
+                                 final Optional<AirbyteSchemaHistoryStorage> schemaHistoryManager) {
+    this.cdcStateHandler = cdcStateHandler;
+    this.targetPosition = targetPosition;
+    this.eventConverter = eventConverter;
+    this.offsetManager = offsetManager;
+    this.schemaHistoryManager = schemaHistoryManager;
+    this.previousCheckpointOffset = (HashMap<String, String>) offsetManager.read();
+    this.initialOffset = new HashMap<>(this.previousCheckpointOffset);
+    resetCheckpointValues();
```

Comment:
nit : not sure if there is a need for this? I understand you're porting over code from before

Reply:
removed. yeah initial value has been empty. Previously it also reset other stuff but I've deleted most of them so it's not applicable here.